### PR TITLE
fix streaming handling for builtin assistants

### PR DESCRIPTION
**Pull request by pmeier**
My patch in #425 is broken when using the Python API with `asyncio.run`, i.e. the regular use case. As implemented in #425, `break`ing out of the loop in `.answer()` means that the async iterator is not fully exhausted, i.e. `AsyncStopIteration` is never raised. This leads to the context manager that handles the HTTP request staying open longer than `Chat.answer`. When it is cleaned up eventually, internally it calls `await request.aclose()`. However, if at that point the async event loop is already killed, e.g. because the coroutine called by `asyncio.run` is finished, you'll get an error:

```
Task exception was never retrieved
future: <Task finished name='Task-5' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('aclose(): asynchronous generator is already running')>
RuntimeError: aclose(): asynchronous generator is already running
Exception ignored in: <coroutine object Response.aclose at 0x73f95efdfcc0>
Traceback (most recent call last):
  File "/home/philip/.conda/envs/ragna-deploy-dev/lib/python3.9/site-packages/httpx/_models.py", line 1008, in aclose
    await self.stream.aclose()
  File "/home/philip/.conda/envs/ragna-deploy-dev/lib/python3.9/site-packages/httpx/_client.py", line 155, in aclose
    await self._stream.aclose()
  File "/home/philip/.conda/envs/ragna-deploy-dev/lib/python3.9/site-packages/httpx/_transports/default.py", line 259, in aclose
    await self._httpcore_stream.aclose()
  File "/home/philip/.conda/envs/ragna-deploy-dev/lib/python3.9/site-packages/httpcore/_async/connection_pool.py", line 374, in aclose
    await self._stream.aclose()
  File "/home/philip/.conda/envs/ragna-deploy-dev/lib/python3.9/site-packages/httpcore/_synchronization.py", line 225, in __exit__
    self._anyio_shield.__exit__(exc_type, exc_value, traceback)
  File "/home/philip/.conda/envs/ragna-deploy-dev/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 406, in __exit__
    if current_task() is not self._host_task:
  File "/home/philip/.conda/envs/ragna-deploy-dev/lib/python3.9/asyncio/tasks.py", line 38, in current_task
    loop = events.get_running_loop()
RuntimeError: no running event loop
```

To fix this, this PR introduces the same pattern for our streaming handling that `httpx` and `httpx_sse` use as well: a context manager creates the stream and the stream can than be iterated on. 

https://github.com/Quansight/ragna/blob/7c5728af3673868222ded6f06bb02f6f6d9cf03a/ragna/assistants/_http_api.py#L83-L89

https://github.com/Quansight/ragna/blob/7c5728af3673868222ded6f06bb02f6f6d9cf03a/ragna/assistants/_http_api.py#L99-L103

Basically, we switch from

```python
async for data in self._call_api(...):
    ...
```

to

```python
async with self._call_api(...) as stream:
    async for data in stream:
        ...
```

With this pattern, the cleanup is correctly handled when we exit `.answer()` and thus get rid of the error.

### CODE REVIEW

**Comment by pmeier** [comment]
> Could you add a unit test to prevent this regression in the future?

I thought about it, but I honestly have trouble coming up with one that is not unreasonably complex. We need a server that we can hit for HTTP streaming. And the best option we have here is to start our own.

I'll see if we can have this in a reasonable way.

File: tests/deploy/ui/test_ui.py [hunk 1]
```diff
@@ -11,12 +10,7 @@
 from ragna._utils import timeout_after
 from ragna.deploy import Config
 from tests.deploy.utils import TestAssistant
-
-
-def get_available_port():
```
Comment (pmeier) [inline tests/deploy/ui/test_ui.py:16]: Moved it to the generic test utils as this is no longer just needed for deploy.

File: tests/utils.py [hunk 2]
```diff
@@ -1,7 +1,27 @@
...
 
 import pytest
 
 skip_on_windows = pytest.mark.skipif(
     platform.system() == "Windows", reason="Test is broken skipped on Windows"
 )
+
+
+@contextlib.contextmanager
+def background_subprocess(*args, stdout=sys.stdout, stderr=sys.stdout, **kwargs):
```
Comment (pmeier) [inline tests/utils.py:15]: We had this before, but it was removed in #322. Probably can also be used by the UI tests, but we can do that in a follow-up.

File: tests/assistants/test_api.py [hunk 3]
```diff
@@ -30,3 +37,105 @@ async def test_api_call_error_smoke(mocker, assistant):
...
+        expected_chunk = next(expected_chunks)
+        assert actual_chunk == expected_chunk
+
+    with pytest.raises(StopIteration):
+        next(expected_chunks)
+
+
+@pytest.mark.parametrize("streaming_protocol", list(HttpStreamingProtocol))
+def test_http_streaming_termination(streaming_server, streaming_protocol):
+    # Non-regression test for https://github.com/Quansight/ragna/pull/462
```
Comment (pmeier) [inline tests/assistants/test_api.py:122]: @nenb This test fails when I revert my patch in this PR. See the CI runs for 0b9211ee366f844570c4edb3ce40e28280951ce5.
